packages/nimble-components/src/menu/menu.foundation.ts

```diff
@@ -0,0 +1,405 @@
+            // set the focus index
+            this.focusIndex = focusIndex;
+        }
+    };
+
+    private readonly handleItemFocus = (e: Event): void => {
+        const targetItem: HTMLElement = e.target as HTMLElement;
+
+        if (
+            this.menuItems !== undefined
+            && targetItem !== this.menuItems[this.focusIndex]
+        ) {
+            this.menuItems[this.focusIndex]?.setAttribute('tabindex', '-1');
+            this.focusIndex = this.menuItems.indexOf(targetItem);
+            targetItem.setAttribute('tabindex', '0');
+        }
+    };
+
+    private readonly handleExpandedChanged = (e: Event): void => {
+        if (
+            e.defaultPrevented
+            || e.target === null
+            || this.menuItems === undefined
+            || !this.menuItems.includes(e.target as Element)
+        ) {
+            return;
+        }
+
+        e.preventDefault();
+        const changedItem: MenuItem = e.target as MenuItem;
+
+        // closing an expanded item without opening another
+        if (
+            this.expandedItem !== null
+            && changedItem === this.expandedItem
+            && changedItem.expanded === false
+        ) {
+            this.expandedItem = null;
+            return;
+        }
+
+        if (changedItem.expanded) {
+            if (
+                this.expandedItem !== null
+                && this.expandedItem !== changedItem
+            ) {
+                this.expandedItem.expanded = false;
+            }
+            this.menuItems[this.focusIndex]?.setAttribute('tabindex', '-1');
+            this.expandedItem = changedItem;
+            this.focusIndex = this.menuItems.indexOf(changedItem);
+            changedItem.setAttribute('tabindex', '0');
+        }
+    };
+
+    private readonly removeItemListeners = (): void => {
+        if (this.menuItems !== undefined) {
+            this.menuItems.forEach((item: Element) => {
+                item.removeEventListener(
+                    'expanded-change',
+                    this.handleExpandedChanged
+                );
+                item.removeEventListener('focus', this.handleItemFocus);
+            });
+        }
+    };
+
+    private itemsChanged(): void {
+        // only update children after the component is connected and
+        // the setItems has run on connectedCallback
+        // (menuItems is undefined until then)
+        if (this.$fastController.isConnected && this.menuItems !== undefined) {
+            this.setItems();
+        }
+    }
+
+    private itemIconsChanged(oldIcons: Element[], newIcons: Element[]): void {
+        // Unintuitively, providing a selector (i.e. ':scope > * > [slot="start"]') to
+        // the children directive does not limit when/how often itemIcons gets updated.
+        // All it does is filter the elements that are assigned to itemIcons.
+        // To avoid unneccessary calls to setItems, we need to check if there was actually a change.
```

Comment:
Useful to know but not specific to this spot, it's general behavior of selector feature of children directive. Think we can leave it off or simplify to something like "detect if slotted icon element instances changed"